packages/firestore/src/core/sync_engine.ts

```diff
@@ -958,9 +964,42 @@ export class SyncEngine implements RemoteSyncer, SharedClientStateSyncer {
           targetId,
           /*current=*/ false
         );
+      } else {
+        // For queries that have a local View, we need to update their state
+        // in LocalStore (as the resume token and the snapshot version
+        // might have changed) and reconcile their views with the persisted
+        // state (the list of syncedDocuments may have gotten out of sync).
+        await this.localStore.releaseTarget(
+          targetId,
+          /*keepPersistedQueryData=*/ true
+        );
+
+        assert(
+          !!this.queriesByTarget[targetId] &&
+            !this.queriesByTarget[targetId].isEmpty(),
+          `There is no queries mapped to target id ${targetId}`
+        );
+
+        const queries = this.queriesByTarget[targetId].toArray();
+        queryData = await this.localStore.allocateTarget(queries[0].toTarget());
```

Comment:
I think this code (and the logic behing it) would be easier to follow if `releaseQuery` and `allocateQuery` would continue to happen right after each other.